core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -178,4 +220,27 @@ private void checkBucketIsAlluxioDirectory(String bucketPath) throws S3Exception
       throw toBucketS3Exception(e, bucketPath);
     }
   }
+
+  private List<URIStatus> listObjects(AlluxioURI uri) throws IOException, AlluxioException {
+    List<URIStatus> objects = new ArrayList<>();
+    Queue<URIStatus> traverseQueue = new ArrayDeque<>();
+    List<URIStatus> children = mFileSystem.listStatus(uri);
+    traverseQueue.addAll(children);
+    while (!traverseQueue.isEmpty()) {
+      URIStatus cur = traverseQueue.remove();
+      if (!cur.isFolder()) {
+        // Alluxio file is an object.
+        objects.add(cur);
+      } else {
+        List<URIStatus> curChildren = mFileSystem.listStatus(new AlluxioURI(cur.getPath()));
+        if (curChildren.isEmpty()) {
+          // An empty Alluxio directory is considered as a valid object.
+          objects.add(cur);
+        } else {
+          traverseQueue.addAll(curChildren);
+        }
+      }
+    }
+    return objects;
```

Comment:
Do we need to return a list of prefixes?

Reply:
good point, I haven't added `commonPrefixes` since `delimiter` is not supported yet. I plan to add in a follow-up PR.